### PR TITLE
✨ Add Heatmap

### PR DESCRIPTION
**Pull request by Vyperus**
**Changes:**

1. Implements a heatmap view for processes.
2. Adds a service and repository for the heatmap.
3. Adds interfaces.
4. Adds events for the inspect, heatmap and dashboard view.
5. Implements an inspect view, where the heatmap or the dashboard is shown.

**Issues:**

Closes #Issue
Closes #Issue
Closes (...)

PR: #810 

---

See the [examples](#example) for inspiration.

## How can others test the changes?

> Describe how others can test your changes (you can remove this section for typo fixes etc.)

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/modules/heatmap/heatmap.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        if (importXmlError) {
+          reject(importXmlError);
+          return;
+        }
+        resolve();
+      });
+    });
+  }
+
+}
```
Comment (cmg-dev) [inline src/modules/heatmap/heatmap.ts]: I made it to here. will continue later.

File: src/modules/heatmap/services/heatmap.service.ts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+      }
+    });
+
+    const participantShape: IShape = this._getParticipantShape(elementRegistry);
+    addOverlay(participantShape.id, participantsTokenCount, {
+      // tslint:disable-next-line:no-magic-numbers
+      left: participantShape.width - 18,
+      // tslint:disable-next-line:no-magic-numbers
+      top: participantShape.height - 20,
+    });
```
[the remark this answers is not in the document]
Comment (alexanderkasten) [reply src/modules/heatmap/services/heatmap.service.ts]: @cmg-dev Der Schritt ist nicht in der forEach und wird nur einmal ausgeführt.

File: src/modules/heatmap/contracts/IFlowNodeAssociation.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export interface IFlowNodeAssociation {
+  associationId: string;
+  sourceId: string;
+  runtime_medianInMs: number;
```
[the remark this answers is not in the document]
Comment (cmg-dev) [reply src/modules/heatmap/contracts/IFlowNodeAssociation.ts]: Bitte dann im 2. Schritt machen.

File: src/modules/heatmap/contracts/constants.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  },
+  events: {
+    left: 23,
+    top: -24,
+  },
+  participants: {
+    left: 18,
+    top: 20,
+  },
+};
```
Comment (cmg-dev) [inline src/modules/heatmap/contracts/constants.ts]: Das können wir besser unter:

`src/modules/heatmap/contracts/position_constants.ts`

File: src/modules/heatmap/heatmap.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
+import {EventAggregator} from 'aurelia-event-aggregator';
+import {bindable, inject} from 'aurelia-framework';
+
+import {IBpmnModeler, IElementRegistry, IOverlay} from '../../contracts/index';
+import {IFlowNodeAssociation, IHeatmapService} from './contracts';
+
+import * as bundle from '@process-engine/bpmn-js-custom-bundle';
+import {ActiveToken, FlowNodeRuntimeInformation} from '@process-engine/kpi_api_contracts';
+import {ProcessModelExecution} from '@process-engine/management_api_contracts';
+import environment from '../../environment';
```
Comment (SteffenKn) [inline src/modules/heatmap/heatmap.ts]: Sort Imports

File: src/modules/heatmap/heatmap.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
+import {EventAggregator} from 'aurelia-event-aggregator';
+import {bindable, inject} from 'aurelia-framework';
+
+import {IBpmnModeler, IElementRegistry, IOverlay} from '../../contracts/index';
+import {IFlowNodeAssociation, IHeatmapService} from './contracts';
```
Comment (SteffenKn) [inline src/modules/heatmap/heatmap.ts]: Sort Imports

File: src/modules/dashboard/dashboard.ts [hunk 7]
```diff
@@ -12,8 +12,7 @@ import {
 } from '@process-engine/management_api_contracts';
 
 import {
-  IAuthenticationService,
-  NotificationType,
+  IAuthenticationService, NotificationType,
 } from '../../contracts/index';
```
Comment (SteffenKn) [inline src/modules/dashboard/dashboard.ts]: import {IAuthenticationService, NotificationType} from '../../contracts/index';

File: src/modules/heatmap/heatmap.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+import environment from '../../environment';
+
+interface RouteParameters {
+  processModelId: string;
+}
+
+@inject('HeatmapService', EventAggregator)
+export class Heatmap {
+  public viewerContainer: HTMLDivElement;
+  @bindable() public processmodelid: string;
```
Comment (SteffenKn) [inline src/modules/heatmap/heatmap.ts]: `processModelId` in html: `process-model-id`

File: src/modules/heatmap/services/heatmap.service.ts [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+            left: position.left,
+            top: position.top,
+          },
+          html: `<div class="heatmap__overlay" title="This element has actual ${count} token.">${countIsTooHigh ? '99+' : count}</div>`,
+        });
+      });
+
+    const elementsForOverlays: Array<IShape> = this._getElementsForOverlays(elementRegistry);
+    const activeTokenListArray: Array<Array<ActiveToken>> = await this._getActiveTokenListArray(elementsForOverlays);
+    this._addShapeTypeToActiveToken(activeTokenListArray, elementsForOverlays);
```
Comment (SteffenKn) [inline src/modules/heatmap/services/heatmap.service.ts]: Add some spacing

File: src/modules/heatmap/services/heatmap.service.ts [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+
+      const elementToColorIsUndefined: boolean = elementToColor === undefined;
+
+      if (elementToColorIsUndefined) {
+        return;
+      }
+
+      const shapeToColor: IShape = this._getShape(elementRegistry, elementToColor);
+
+      if (elementToColor.medianRuntimeInMs > association.runtime_medianInMs) {
```
Comment (SteffenKn) [inline src/modules/heatmap/services/heatmap.service.ts]: Extract Business Rule

File: src/modules/heatmap/services/heatmap.service.ts [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+  private _getElementsToColor(
+    associations: Array<IFlowNodeAssociation>,
+    flowNodeRuntimeInformation: Array<FlowNodeRuntimeInformation>,
+  ): Array<FlowNodeRuntimeInformation> {
+
+    const elementsToColor: Array<FlowNodeRuntimeInformation> = flowNodeRuntimeInformation.filter((information: FlowNodeRuntimeInformation) => {
+      const associationWithSameId: IFlowNodeAssociation = associations.find((association: IFlowNodeAssociation) => {
+        return association.sourceId === information.flowNodeId;
+      });
+      return associationWithSameId;
```
Comment (SteffenKn) [inline src/modules/heatmap/services/heatmap.service.ts]: Add some spacing

File: src/modules/heatmap/services/heatmap.service.ts [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+    });
+  }
+
+  private _getElementsWithoutToken(elementsForOverlays: Array<IShape>, activeTokenListArray: Array<Array<ActiveToken>>): Array<IShape> {
+    const elementsWithoutToken: Array<IShape> = elementsForOverlays.filter((element: IShape) => {
+      const activeTokenForElement: Array<ActiveToken> = activeTokenListArray.find((activeTokenArray: Array<ActiveToken>) => {
+        return activeTokenArray[0].flowNodeId === element.id;
+      });
+
+      if (activeTokenForElement === undefined) {
```
Comment (SteffenKn) [inline src/modules/heatmap/services/heatmap.service.ts]:
```
const noActiveTokenForElement: boolean = activeTokenForElement === undefined;

return noActiveTokenForElement;
```

File: src/modules/inspect/inspect.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
+import {EventAggregator} from 'aurelia-event-aggregator';
+import {bindable, inject} from 'aurelia-framework';
+
+import {Dashboard} from '../dashboard/dashboard';
+import {Heatmap} from '../heatmap/heatmap';
+
+import {IDiagram} from '@process-engine/solutionexplorer.contracts';
+import environment from '../../environment';
```
Comment (SteffenKn) [inline src/modules/inspect/inspect.ts]: Sort Imports

File: src/modules/inspect/inspect.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+    this.processModelId = routeParameters.processModelId;
+    const process: IDiagram = {
+      name: this.processModelId,
+      xml: '',
+      uri: '',
+      id: this.processModelId,
+    };
+
+    if (routeParameters.view === 'dashboard') {
```
Comment (SteffenKn) [inline src/modules/inspect/inspect.ts]: Extract Business Rule

File: src/modules/inspect/inspect.ts [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+      setTimeout(() => {
+        const dashboardIsAttached: boolean = this.dashboard !== undefined;
+
+        if (dashboardIsAttached) {
+          this.dashboard.canActivate();
+        }
+      }, 0);
+
+      if (routeParameters.latestSource === 'process-engine') {
```
Comment (SteffenKn) [inline src/modules/inspect/inspect.ts]: Extract Business Rule

File: src/modules/inspect/inspect.ts [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+        this._eventAggregator.publish(environment.events.navBar.showInspectButtons);
+        this._eventAggregator.publish(environment.events.navBar.disableDashboardAndEnableHeatmapButton);
+        this._eventAggregator.publish(environment.events.navBar.showProcessName, process);
+      }
+    } else if (routeParameters.view === 'heatmap') {
```
Comment (SteffenKn) [inline src/modules/inspect/inspect.ts]: Extract Business Rule

File: src/modules/navbar/navbar.ts [hunk 17]
```diff
@@ -51,9 +65,17 @@ export class NavBar {
 
       this._eventAggregator.subscribe(environment.events.navBar.updateProcess, (process: IDiagram) => {
         this.process = process;
+
+        const processIdIsUndefined: boolean = process.id === undefined;
+        processIdIsUndefined ? this.latestSource = 'file-system' : this.latestSource = 'process-engine';
```
Comment (SteffenKn) [inline src/modules/navbar/navbar.ts]: ```this.latestSource =  processIdIsUndefined ? 'file-system' : 'process-engine';```

File: src/modules/navbar/navbar.ts [hunk 18]
```diff
@@ -102,6 +150,78 @@ export class NavBar {
...
+      latestSource: undefined,
+    });
+  }
+
+  public navigateToDesigner(): void {
+    const processIsUndefined: boolean = this.process === undefined;
+
+    if (processIsUndefined) {
+      this._notificationService.showNotification(NotificationType.INFO, 'To open the designer, you must first select a diagram');
+      return;
```
Comment (SteffenKn) [inline src/modules/navbar/navbar.ts]: Add some spacing

File: src/modules/navbar/navbar.ts [hunk 19]
```diff
@@ -102,6 +150,78 @@ export class NavBar {
...
+    }
+
+    if (this.latestSource === 'process-engine') {
```
Comment (SteffenKn) [inline src/modules/navbar/navbar.ts]: Extract business rule

File: src/modules/navbar/navbar.ts [hunk 20]
```diff
@@ -102,6 +150,78 @@ export class NavBar {
...
+      this._router.navigateToRoute('processdef-detail', {
+        processModelId: this.process.id,
+      });
+    } else if (this.latestSource === 'file-system') {
```
Comment (SteffenKn) [inline src/modules/navbar/navbar.ts]: Extract business rule

File: src/modules/process-list/process-list.ts [hunk 21]
```diff
@@ -49,9 +49,11 @@ export class ProcessList {
     this._authenticationService = authenticationService;
   }
 
-  public currentPageChanged(newValue: number, oldValue: number): void {
+  public async currentPageChanged(newValue: number, oldValue: number): Promise<void> {
     if (oldValue !== undefined && oldValue !== null) {
```
Comment (SteffenKn) [inline src/modules/process-list/process-list.ts]: Extract business rule

File: src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts [hunk 22]
```diff
@@ -94,13 +97,26 @@ export class SolutionExplorerSolution {
   }
 
   public attached(): void {
+    this._subscriptions = [
+      this._eventAggregator.subscribe(environment.events.processSolutionPanel.navigateToHeatmap, (heatmapView?: string) => {
+        this._diagramRoute = 'inspect';
+        const heatmapViewIsDashboard: boolean = heatmapView === 'dashboard';
+        heatmapViewIsDashboard ? this._inspectView = 'dashboard' : this._inspectView = 'heatmap';
```
Comment (SteffenKn) [inline src/modules/solution-explorer/solution-explorer-solution/solution-explorer-solution.ts]: Add some spacing.

Also:
```
        this._inspectView = heatmapViewIsDashboard ? 'dashboard' : 'heatmap';
```

So it should look like this:
```
        const heatmapViewIsDashboard: boolean = heatmapView === 'dashboard';

        this._diagramRoute = 'inspect';
        this._inspectView = heatmapViewIsDashboard ? 'dashboard' : 'heatmap';
```

File: src/modules/process-solution-panel/process-solution-panel.html [hunk 23]
```diff
@@ -0,0 +1,83 @@
+<template>
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.html]: Delete this file

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 24]
```diff
@@ -0,0 +1,635 @@
+import {EventAggregator, Subscription} from 'aurelia-event-aggregator';
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.ts]: Delete this file.

File: src/modules/navbar/navbar.ts [hunk 25]
```diff
@@ -102,6 +150,78 @@ export class NavBar {
...
+
+    const activeRouteIsInspect: boolean = this.activeRouteName === 'inspect';
+
+    if (activeRouteIsInspect) {
+      return;
+    }
+
+    const activeRouteIsNotProcessEngineRoute: boolean = this.activeRouteName !== 'processdef-detail';
+
+    let processModelId: string = this.process ? this.process.id : undefined;
```
Comment (robpal94) [inline src/modules/navbar/navbar.ts]: You could avoid `let` here

```
const processModelId: string = (this.process && !activeRouteIsNotProcessEngineRoute) 
     ? this.process.id
     : undefined;
```

File: src/modules/navbar/navbar.ts [hunk 26]
```diff
@@ -102,6 +150,78 @@ export class NavBar {
...
+      view: this.inspectView,
+      latestSource: undefined,
+    });
+  }
+
+  public navigateToDesigner(): void {
+    const processIsUndefined: boolean = this.process === undefined;
+
+    if (processIsUndefined) {
+      this._notificationService.showNotification(NotificationType.INFO, 'To open the designer, you must first select a diagram');
```
Comment (robpal94) [inline src/modules/navbar/navbar.ts]: _In order to open the designer, you have to select a diagram first!_

File: src/modules/heatmap/services/heatmap.service.ts (hunk 10 above)
Comment (robpal94) [reply src/modules/heatmap/services/heatmap.service.ts]: We also should unify the name `medianRuntimeInMS`.

File: src/modules/heatmap/heatmap.scss [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+.heatmap__viewer-container {
+  height: 100%;
+}
+
+.heatmap__legend {
+  position: absolute;
+  top: 60px;
+  right: 56px;
+  max-width: 155px;
+  background: whitesmoke;
```
[the remark this answers is not in the document]
Comment (SteffenKn) [reply src/modules/heatmap/heatmap.scss]: yes

**Pull request merged**